src/AdoNetCore.AseClient/Internal/TokenStream.cs

```diff
@@ -0,0 +1,260 @@
+                    break;
+                }
+            }
+
+            // If cancelled, burn
+            if (IsCancelled)
+            {
+                BurnPackets();
+            }
+
+            return bytesWrittenToBuffer;
+        }
+
+        /// <summary>
+        /// It is possible that the caller doesn't process all of the data returned to the socket.
+        /// This method reads and discards all data, leaving the socket in a ready state for the next request.
+        /// </summary>
+        private void BurnPackets()
+        {
+            while (_networkHasBytes)
+            {
+                BufferBytes(_headerBuffer, 0, _environment.HeaderSize);
+
+                var bufferStatus = (BufferStatus) _headerBuffer[1];
+
+                var length = _headerBuffer[2] << 8 | _headerBuffer[3];
+
+                _bodyBufferLength = length - _environment.HeaderSize;
+                _bodyBufferPosition = 0;
+
+                BufferBytes(_bodyBuffer, 0, _bodyBufferLength);
+
+                // If there is no more data, stop there.
+                if (bufferStatus.HasFlag(BufferStatus.TDS_BUFSTAT_EOM) || length == _environment.HeaderSize)
+                {
+                    _networkHasBytes = false;
+                    break;
+                }
+            }
+
+            _bufferHasBytes = false;
+            _bodyBufferLength = 0;
+            _bodyBufferPosition = 0;
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotImplementedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!_isDisposed)
+            {
+                _isDisposed = true;
+
+                if (!disposing)
+                {
+                    BurnPackets(); // Considering the case where the socket lives longer than the stream, and the stream is disposed without reading all of the data.
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        /// <summary>
+        /// Ensure that the requested bytes have been received. The base <see cref="NetworkStream"/> will
+        /// block until at least one byte is available, but not necessarily wait for all of the requested
+        /// bytes.
+        /// </summary>
+        /// <param name="buffer">The buffer to fill.</param>
+        /// <param name="offset">The position within the buffer to write data into.</param>
+        /// <param name="count">The number of bytes.</param>
+        private void BufferBytes(byte[] buffer, int offset, int count)
+        {
+            // TODO - this should not just loop, burning CPU.
+            if (count > 0)
+            {
+                var remainingBytes = count;
+                var totalReceivedBytes = 0;
+                do
+                {
+                    var receivedBytes = base.Read(buffer, offset + totalReceivedBytes, remainingBytes);
```

Comment:
One thing I've noticed in other development is that a `socket.Read` can return `0` if the connection breaks (remote closes the connection), rather than throwing a SocketException. *If* that happens here, you get an infinite loop -- might be good to look into it.